### ISSUE TITLE
Gas price engine API

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/web3.py#494]
### What was wrong?

Currently, web3.py just uses the `eth_gasPrice` JSON-RPC API to determine the gas price to use for transactions.  Most nodes do not currently implement very good algorithms for determining gas prices.

### How can it be fixed?

Lets implement a new API for determining gas prices.

* Add a new API: `Web3.set_gas_price_engine(engine)`
* Add new API: `Web3.get_gas_price(transaction_params)` which delegates to whatever *engine* is set.

An `engine` must be a callable with the function signature:

```
engine(w3, transaction_params)
```

* `w3` will be the `Web3` instance
* `transaction_params` will be a dictionary of the transaction params being sent.

The callable **must** return a positive integer which represents the `gas_price` in the *wei* denomination.

The **default** engine that a web3 instance should use is one that simply returns `web3.eth.gasPrice`

Documentation should be added covering the following topics.

* Overview of the `Web3.set_gas_price_engine()` API (maybe in `docs/overview.rst`)
* Writing your own gas price engine (maybe in new document `docs/gas_pricing_engine.rst` or `docs/web3.main.rst`)
* List of available engines, where to import them and how to use them.

**Comment by carver** [comment]
Ethgasstation is pretty good, but during high rates of change it can actually be pretty far off. It's actually quite a tricky problem, because there are a lot of hidden variables that could be influencing which transactions are included. It may be intractable to get very close to the "right" answer.

So instead of this (or perhaps in addition to it), another good solution might be to monitor pending transactions in a daemon and bump the gas price up over time if they don't get included.

**Comment by carver** [comment]
That's sufficiently different that it belongs in another issue. Or better, in its own library. Another thing on the todo list...

**Comment by pipermerriam** [comment]
> during high rates of change it can actually be pretty far off.

do you know what the sample size that Eth Gas Station uses is?

> Or better, in its own library

I'm assuming you're talking about the actual EGS algorithm?  Maybe but I'm not sure what it's really useful for on it's own..  I could probably be convinced either way

**Comment by carver** [comment]
>> JC: Or better, in its own library
> 
> PM: I'm assuming you're talking about the actual EGS algorithm? Maybe but I'm not sure what it's really useful for on it's own.. I could probably be convinced either way

I meant to say that the following should go in its own library:

> JC: to monitor pending transactions in a daemon and bump the gas price up over time if they don't get included.

**Comment by gitcoinbot** [comment]
__This issue now has a funding of 0.18 ETH (121.5 USD) attached to it.__

 * If you would like to work on this issue you can claim it [here](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/494).
 * If you've completed this issue and want to claim the bounty you can do so [here](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/494)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $9757.26 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by monokh** [comment]
This is the script ethgasstation uses fyi: 
https://github.com/ethgasstation/ethgasstation-adaptive-oracle
What i always wondered about was if there was a simple and straightforward approach for the libraries like web3 to get going with. Something like get the latest block and pick out median/mode/avg of the gas prices in that block. While it isn't the most economical/accurate, it should be fine in most cases.
It feels like that there would be quite a substantial effort to get very accurate gas prices - as is visible from the ethgasstation implementation.

**Comment by pipermerriam** [comment]
@monokh take a look at #496 which builds on top of the API suggested in this PR and is a simplified version of the EGS algorithm.

**Comment by gitcoinbot** [comment]
__The funding of 0.18 ETH (119.91 USD) attached has been claimed by @monokh.__ 

 @monokh, please leave a comment to let the funder (@owocki) and the other parties involved your implementation plan.  If you don't leave a comment, the funder may expire your claim at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/494)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $10235.35 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by owocki** [comment]
@monokh please let us know how its going.

**Comment by monokh** [comment]
@owocki going well 😃  _Lots_ of discussion on the PR but we are quite close to resolution I think.

**Comment by pipermerriam** [comment]
@owocki we can consider this closed/complete.  I'll match the current bounty, let me know how to do that.

**Comment by owocki** [comment]
@pipermerriam remitting the bounty now. 

> I'll match the current bounty, let me know how to do that.

@monokh 's ETH address is listed at 
https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/494 -- you can send ETH direct to him there.. OR you can tip any github username at https://gitcoin.co/tip

**Comment by gitcoinbot** [comment]
__The funding of 0.18 ETH (101.07 USD) attached to this issue has been approved & issued to @monokh.__  

 * Learn more at [on the gitcoin issue page](https://gitcoin.co/funding/details?url=https://github.com/ethereum/web3.py/issues/494)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $8751.07 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by monokh** [comment]
Thanks folks 😃  really enjoyed this one
@pipermerriam 
#496 unblocked now i guess. Going to be really cool seeing that in action!

**Comment by pipermerriam** [comment]
@monokh etherscan.io/tx/0x20718ee637c2c01dfab0c7c0e9747f50512ef934e94462f077ced545d8926683
